lib/libc/semaphore/sem_init.c

```diff
@@ -60,11 +60,39 @@
 #include <limits.h>
 #include <semaphore.h>
 #include <errno.h>
+#if defined(CONFIG_BINMGR_RECOVERY) && defined(__KERNEL__)
+#include <queue.h>
+#include <tinyara/arch.h>
+#endif
 
 #ifdef CONFIG_SEMAPHORE_HISTORY
 #include <tinyara/debug/sysdbg.h>
 #endif
 
+#if defined(CONFIG_BINMGR_RECOVERY) && defined(__KERNEL__)
+sq_queue_t g_sem_list;
+
+static void sem_register(FAR sem_t *sem)
```

Comment:
How about separating register and unregister to another file? Because, sem_init is in lib, but sem_register is only for kernel.